File: cpp/structural/flyweight.cpp
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by GNU General Public License
// that can be found in the LICENSE file.

#include <iostream>
#include <string>
#include <map>
#include <utility>

class KarakTea {

};

class TeaMaker {
 public:
  KarakTea make(const std::string& preference) {
    if (available_tea_.find(preference) == available_tea_.cend()) {
      available_tea_.emplace(preference, KarakTea{});
    }

    return available_tea_.at(preference);
  }

 private:
  std::map<std::string, KarakTea> available_tea_;
};

class TeaShop {
 public:
  explicit TeaShop(const TeaMaker& maker) : maker_(maker), orders_() {}

  void take_order(const std::string& tea_type, int table) {
    orders_[table] = maker_.make(tea_type);
  }
  void serve() {
    for (const auto& item : orders_) {
      std::cout << "Serving tea to table: " << item.first << std::endl;
    }
  }

 private:
  TeaMaker maker_;
  std::map<int, KarakTea> orders_;
};

int main() {
  TeaMaker maker;
  TeaShop shop(maker);
  shop.take_order("less sugar", 1);
  shop.take_order("more milk", 2);
  shop.take_order("without sugar", 3);

  shop.serve();

  return 0;
}

File: code/design_patterns/structural/src/bin/flyweight.rs
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be
// found in the LICENSE file.

#![deny(
    warnings,
    clippy::all,
    clippy::cargo,
    clippy::nursery,
    clippy::pedantic
)]

use std::collections::BTreeMap;

#[derive(Debug, Default, Clone)]
pub struct KarakTea {
    preference: String,
}

impl KarakTea {
    #[must_use]
    pub const fn new(preference: String) -> Self {
        Self { preference }
    }

    #[must_use]
    pub fn preference(&self) -> &str {
        self.preference.as_str()
    }
}

#[derive(Default)]
pub struct TeaMaker {
    available_tea: BTreeMap<String, KarakTea>,
}

impl TeaMaker {
    #[must_use]
    #[allow(clippy::missing_panics_doc)]
    pub fn make(&mut self, preference: &str) -> KarakTea {
        if let Some(tea) = self.available_tea.get(preference) {
            tea.clone()
        } else {
            let tea = KarakTea::new(preference.to_owned());
            self.available_tea
                .insert(preference.to_owned(), tea.clone());
            tea
        }
    }
}

pub struct TeaShop {
    tea_maker: TeaMaker,
    orders: BTreeMap<usize, KarakTea>,
}

impl TeaShop {
    #[must_use]
    pub const fn new(tea_maker: TeaMaker) -> Self {
        Self {
            tea_maker,
            orders: BTreeMap::new(),
        }
    }

    pub fn take_order(&mut self, tea_preference: &str, table: usize) {
        let tea = self.tea_maker.make(tea_preference);
        self.orders.insert(table, tea);
    }

    pub fn serve(&self) {
        for (table, tea) in &self.orders {
            println!("Serving the tea( {} ) to table {table}", tea.preference());
        }
    }
}

fn main() {
    let tea_maker = TeaMaker::default();
    let mut tea_shop = TeaShop::new(tea_maker);

    tea_shop.take_order("less sugar", 1);
    tea_shop.take_order("more milk", 2);
    tea_shop.take_order("without sugar", 3);
    tea_shop.serve();
}
